server/server_session.cpp

```diff
@@ -6,110 +6,106 @@
-                timer_.expires_after(std::chrono::seconds(TIMEOUT));
                 // process request and send out response
                 std::string response = processor.dispatch(request_str);
-                ws_.async_write(boost::asio::buffer(response), yield);
+                // current session writes doesn't have a name, empty strings are also cheap
+                write({}, std::move(response));
                 BOOST_LOG_SEV(logger_, sev::info)
                     << "request processed in " << stopwatch.elapsed_ms() << "ms";
             }
         }
         // unrecoverable error
         catch (std::exception &e) {
-            // moving this logging line to the end will cause crashes
             BOOST_LOG_SEV(logger_, sev::info) << "Session ending, reason: " << e.what();
-            if (processor.get_session().userid) {
-                db_.logout(*processor.get_session().userid, yield);
-            }
-            close_down_ = true;
-            boost::system::error_code ec;
-            timer_.cancel(ec);
+        }
+        if (processor.get_session().player) {
+            ss_map_.remove(processor.get_session().player.value().name);
+        }
+        if (processor.get_session().userid) {
+            db_.logout(processor.get_session().userid.value(), yield);
         }
     });
+}
 
-    auto self2(shared_from_this());
-    boost::asio::spawn(strand_, [this, self2](boost::asio::yield_context yield) {
-        boost::system::error_code ec;
-        try {
-            for (;;) {
-                if (close_down_)
-                    break;
-                // check it has expired
-                if (timer_.expiry() <= std::chrono::steady_clock::now()) {
-                    if (ws_.is_open() && ping_state_ == pingstate::responsive) {
-                        ping_state_ = pingstate::onhold;
-                        timer_.expires_after(std::chrono::seconds(TIMEOUT));
-                        // Now send the ping
-                        ws_.async_ping({}, yield[ec]);
-                        // we don't care about ec here
-                    } else if (ws_.is_open()) {
-                        // ping state is onhold - no activity in last 15 seconds
-                        BOOST_LOG_SEV(logger_, sev::info)
-                            << "Connection closing due to ping timeout";
-                        ws_.next_layer().next_layer().cancel(ec);
-                        break;
-                    }
-                }
-                // wait on the timer
-                timer_.async_wait(yield[ec]);
-                // we need this != check, need to experiment more
-                if (ec && ec != boost::asio::error::operation_aborted)
-                    break;
+void server_session::write(std::string name, std::string str) {
+    BOOST_LOG_SEV(logger_, boost::log::trivial::debug) << "write called";
+    boost::asio::spawn(strand_, [this, name{std::move(name)}, str{std::move(str)},
```

Comment:
why `name{std::move(name)}`? Not just `name`?

Reply:
by default `name` is capture by value, in other words it will make a copy. `move` helps with efficiency